src/library/scala/collection/ArrayOps.scala

```diff
@@ -104,24 +104,25 @@ object ArrayOps {
     def withFilter(q: A => Boolean): WithFilter[A] = new WithFilter[A](a => p(a) && q(a), xs)
   }
 
-  private class ArrayIterator[A](private[this] val xs: Array[A]) extends AbstractIterator[A] {
+  private final class ArrayIterator[@specialized(AnyRef, Int, Double, Long, Float, Char, Byte, Short, Boolean, Unit) A](xs: Array[A]) extends AbstractIterator[A] {
     private[this] var pos = 0
-    def hasNext: Boolean = pos < xs.length
+    private[this] val len = xs.length
+    def hasNext: Boolean = pos < len
     def next(): A = try {
```

Comment:
is is faster to do the explicit range check rather than try/catch

Reply:
Given that the range check will only be needed for those who call next() without a preceding hasNext, I think that using try-catch in this case is rather clever.